### PR TITLE
Switch from Makefile to npm scripts

### LINKED ISSUE #260: Convert make tasks to be npm scripts
Currently, we have make to compile, run tests etc. This doesn't work on Windows ([without installing make](https://www.google.cl/search?q=make%20for%20windows)).
npm scripts are more than capable of performing the same tasks without the dependency.

### PR DESCRIPTION
**Pull request by winniehell**
Closes #260

### CODE REVIEW

**Comment by winniehell** [comment]
I am not sure how we can preserve the V8 options here. That is why I also dropped the failing 0.12 from the Travis CI config.

File: package.json [hunk 1]
```diff
@@ -9,7 +9,15 @@
     "_metalsmith": "bin/_metalsmith"
   },
   "scripts": {
-    "test": "make test"
+    "postinstall": "touch node_modules",
```
Comment (winniehell) [inline package.json]: I don't know why we need this, but I kept it for now.

Comment (woodyrew) [inline package.json]: Not sure why that's there either, I think you can remove it.

File: package.json [hunk 2]
```diff
@@ -9,7 +9,15 @@
...
+    "pretest": "npm install",
+    "test": "_mocha",
```
Comment (woodyrew) [inline package.json]: The following will fix `0.12` in the short term which we need. We're moving to Babel so this won't be needed for long.
    "node_flags": "node --v8-options | grep generators | cut -d ' ' -f 3",
    "test": "node `npm run -s node_flags` ./node_modules/.bin/_mocha"

_Note_: This won't work for Windows but `npm run test-debug` will work in the short term if needed.

File: .travis.yml [hunk 3]
```diff
@@ -2,7 +2,6 @@ language: node_js
 sudo: false
 
 node_js:
-  - "0.12"
```
Comment (woodyrew) [inline .travis.yml]: Don't forget to add this again. (See other comments)

**Comment by winniehell** [comment]
@woodyrew Thank you for the review! :+1: I think I found a simpler way to enable v0.12. What do you think?

File: .travis.yml [hunk 4]
```diff
@@ -13,5 +13,8 @@ matrix:
   allow_failures:
     - node_js: "node"
 
+before_script:
+  - export MOCHA_OPTS=$(node --v8-options | grep --only-matching -- '--harmony_generators')
+
 after_script:
-  make coveralls
+  npm run coveralls
```
Comment (Ajedi32) [inline .travis.yml]: I don't think these commands are equivalent. The Makefile was set to run Istanbul first, whereas this just runs coveralls.

File: package.json [hunk 5]
```diff
@@ -9,7 +9,14 @@
     "_metalsmith": "bin/_metalsmith"
   },
   "scripts": {
-    "test": "make test"
+    "pretest": "npm install",
+    "test": "mocha $MOCHA_OPTS",
+    "pretest-debug": "npm install",
+    "test-debug": "mocha debug",
+    "precoverage": "npm install",
+    "coverage": "istanbul cover ./node_modules/.bin/_mocha",
```
Comment (Ajedi32) [inline package.json]: Looks like Istanbul isn't getting run with the generator flags on Node 0.12. Travis is erroring during this step: https://travis-ci.org/metalsmith/metalsmith/jobs/171375235

Comment (winniehell) [inline package.json]: @Ajedi32 yep, just realized that myself. The problem here is `--harmony_generators` is only supported by `mocha` while istanbul needs to be called with `_mocha`.

But shouldn't it actually be sufficient to run the coverage with only one node version?

Comment (winniehell) [inline package.json]: Maybe the coverage part should actually go into the [`deploy` section] with a restriction to one node version.

[`deploy` section]: https://docs.travis-ci.com/user/deployment/

Comment (Ajedi32) [inline package.json]: The Makefile worked by running Istanbul like this: `node --harmony_generators ./node_modules/.bin/istanbul`. Couldn't you just do that?

**Comment by winniehell** [comment]
@Ajedi32 @woodyrew Looks like it's fixed now. :smiley:

**Pull request merged**